dateutil/parser/_parser.py

```diff
@@ -981,6 +1049,15 @@ def _parse_numeric_token(self, tokens, idx, info, ymd, res, fuzzy):
                 idx += 1
             idx += 1
 
+        elif (not ymd.has_year and len_li == 4 and
+              info.month(next_jump(tokens, idx)[0] or '')):
+            # e.g. this is the "0003" in "0003 Nov 06"
+            # unambiguous year, use value_repr instead of value to avoid
+            #  dropping leading zeros, e.g. "0031" represents 31 AD,
+            #  not 2031
+            # Note: this check must come before the check for info.jump below
+            ymd.append(value_repr, 'Y')
```

Comment:
I'm mildly surprised that we can put strings in the `ymd` tuple like this without it breaking anything.

Reply:
_ymd.append starts with a check for `hasattr(val, '__len__')` which I guess is intended to catch strings (not sure why it doesnt just check for strings directly...)